File: apps/web/src/lib/api/work-items.ts
import { apiFetch } from '@/lib/api/client';

export const WORK_ITEM_TYPES = [
  'task',
  'call',
  'meeting',
  'follow_up',
  'site_visit',
  'document_request',
  'proposal_follow_up',
  'reservation_follow_up',
  'deposit_follow_up',
  'contract_follow_up',
  'closing_follow_up',
  'other',
] as const;

export const WORK_ITEM_STATUSES = [
  'open',
  'in_progress',
  'waiting',
  'blocked',
  'completed',
  'cancelled',
  'overdue',
  'archived',
] as const;

export const WORK_ITEM_PRIORITIES = ['low', 'medium', 'high', 'urgent'] as const;

export type WorkItemType = (typeof WORK_ITEM_TYPES)[number];
export type WorkItemStatus = (typeof WORK_ITEM_STATUSES)[number];
export type WorkItemPriority = (typeof WORK_ITEM_PRIORITIES)[number];

export type WorkViewName =
  | 'today'
  | 'overdue'
  | 'upcoming'
  | 'meetings'
  | 'calls'
  | 'follow_ups'
  | 'waiting'
  | 'blocked'
  | 'completed'
  | 'my_work'
  | 'team_work';

export interface MeetingRecord {
  id: string;
  work_item_id: string;
  meeting_type: string;
  location: string | null;
  meeting_url: string | null;
  agenda: string | null;
  notes: string | null;
  outcome: string | null;
  decision_summary: string | null;
  next_steps: string | null;
  started_at: string | null;
  ended_at: string | null;
}

export interface FollowUpRecord {
  id: string;
  work_item_id: string;
  follow_up_type: WorkItemType;
  contact_method: string;
  outcome: string | null;
  response_status: string;
  next_follow_up_at: string | null;
  notes: string | null;
}

export interface WorkItem {
  id: string;
  title: string;
  description: string | null;
  work_item_type: WorkItemType;
  status: WorkItemStatus;
  effective_status: WorkItemStatus | null;
  priority: WorkItemPriority;
  assigned_user_id: string | null;
  created_by_user_id: string | null;
  due_at: string | null;
  start_at: string | null;
  completed_at: string | null;
  cancelled_at: string | null;
  reminder_at: string | null;
  lead_id: string | null;
  opportunity_id: string | null;
  party_id: string | null;
  project_id: string | null;
  inventory_asset_id: string | null;
  proposal_id: string | null;
  outcome: string | null;
  is_private: boolean;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
  meeting: MeetingRecord | null;
  follow_up: FollowUpRecord | null;
}

export interface WorkItemListResponse {
  items: WorkItem[];
  total: number;
  offset: number;
  limit: number;
}

export interface WorkDashboardKpis {
  today_count: number;
  overdue_count: number;
  meetings_today_count: number;
  blocked_count: number;
  waiting_count: number;
  my_work_count: number;
  no_next_action_count: number;
}

export interface CalendarEvent {
  id: string;
  title: string;
  work_item_type: WorkItemType;
  status: WorkItemStatus;
  start_at: string | null;
  due_at: string | null;
  assigned_user_id: string | null;
  is_private: boolean;
}

export interface WorkItemInput {
  title: string;
  description?: string | null;
  work_item_type?: WorkItemType;
  priority?: WorkItemPriority;
  assigned_user_id?: string | null;
  due_at?: string | null;
  start_at?: string | null;
  lead_id?: string | null;
  opportunity_id?: string | null;
  party_id?: string | null;
  proposal_id?: string | null;
  is_private?: boolean;
}

export async function fetchWorkDashboardKpis(): Promise<WorkDashboardKpis> {
  return apiFetch<WorkDashboardKpis>('/sales/work/dashboard/kpis');
}

export async function fetchWorkView(view: WorkViewName, limit = 100): Promise<WorkItemListResponse> {
  return apiFetch<WorkItemListResponse>(`/sales/work/views/${view}?limit=${limit}`);
}

export async function fetchWorkItems(params: Record<string, string | number | boolean | undefined> = {}): Promise<WorkItemListResponse> {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') search.set(key, String(value));
  });
  const qs = search.toString();
  return apiFetch<WorkItemListResponse>(`/sales/work/items${qs ? `?${qs}` : ''}`);
}

export async function fetchWorkItem(id: string): Promise<WorkItem> {
  return apiFetch<WorkItem>(`/sales/work/items/${id}`);
}

export async function createWorkItem(payload: WorkItemInput): Promise<WorkItem> {
  return apiFetch<WorkItem>('/sales/work/items', { method: 'POST', body: JSON.stringify(payload) });
}

export async function updateWorkItem(id: string, payload: Partial<WorkItemInput>): Promise<WorkItem> {
  return apiFetch<WorkItem>(`/sales/work/items/${id}`, { method: 'PATCH', body: JSON.stringify(payload) });
}

export async function completeWorkItem(id: string, payload: { outcome?: string; next_follow_up?: Record<string, unknown> } = {}): Promise<WorkItem> {
  return apiFetch<WorkItem>(`/sales/work/items/${id}/complete`, { method: 'POST', body: JSON.stringify(payload) });
}

export async function cancelWorkItem(id: string, reason?: string): Promise<WorkItem> {
  return apiFetch<WorkItem>(`/sales/work/items/${id}/cancel`, {
    method: 'POST',
    body: JSON.stringify({ status: 'cancelled', reason }),
  });
}

export async function rescheduleWorkItem(id: string, due_at: string): Promise<WorkItem> {
  return apiFetch<WorkItem>(`/sales/work/items/${id}/reschedule`, {
    method: 'POST',
    body: JSON.stringify({ due_at }),
  });
}

export async function createMeeting(payload: WorkItemInput & { meeting_type?: string; meeting_url?: string; agenda?: string }): Promise<WorkItem> {
  return apiFetch<WorkItem>('/sales/work/meetings', { method: 'POST', body: JSON.stringify(payload) });
}

export async function createFollowUp(payload: WorkItemInput & { contact_method?: string }): Promise<WorkItem> {
  return apiFetch<WorkItem>('/sales/work/follow-ups', { method: 'POST', body: JSON.stringify(payload) });
}

export async function fetchCalendarEvents(from_at?: string, to_at?: string): Promise<CalendarEvent[]> {
  const params = new URLSearchParams();
  if (from_at) params.set('from_at', from_at);
  if (to_at) params.set('to_at', to_at);
  const qs = params.toString();
  return apiFetch<CalendarEvent[]>(`/sales/work/calendar${qs ? `?${qs}` : ''}`);
}

export async function fetchLeadWorkItems(leadId: string): Promise<WorkItemListResponse> {
  return apiFetch<WorkItemListResponse>(`/sales/work/leads/${leadId}/items`);
}

export async function fetchOpportunityWorkItems(opportunityId: string): Promise<WorkItemListResponse> {
  return apiFetch<WorkItemListResponse>(`/sales/work/opportunities/${opportunityId}/items`);
}

File: apps/api/src/investhome_api/api/routes/work_items.py
"""Work item API routes."""

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from investhome_api.api.deps.auth import require_permission
from investhome_api.db.session import get_db
from investhome_api.models.work_item import FollowUpRecord, MeetingRecord, WorkItem, WorkItemParticipant
from investhome_api.models.user_auth import User
from investhome_api.models.work_item import WorkItemPriority, WorkItemStatus, WorkItemType
from investhome_api.schemas.work_item import (
    CompleteWorkItemRequest,
    FollowUpCompleteRequest,
    FollowUpCreate,
    MeetingCompleteRequest,
    MeetingCreate,
    MeetingUpdate,
    RescheduleRequest,
    StatusChangeRequest,
    WorkDashboardKpisResponse,
    WorkItemCreate,
    WorkItemListResponse,
    WorkItemResponse,
    WorkItemUpdate,
    CalendarEventResponse,
)
from investhome_api.services.work import work_item_service as svc

router = APIRouter(prefix="/sales/work", tags=["work-items"])


def _handle_error(exc: svc.WorkItemError) -> HTTPException:
    return HTTPException(status_code=exc.status_code, detail=exc.error_key)


def _enrich(db: Session, item: WorkItem) -> WorkItemResponse:
    meeting = db.scalar(select(MeetingRecord).where(MeetingRecord.work_item_id == item.id))
    follow_up = db.scalar(select(FollowUpRecord).where(FollowUpRecord.work_item_id == item.id))
    participants = list(
        db.scalars(select(WorkItemParticipant).where(WorkItemParticipant.work_item_id == item.id)).all()
    )
    response = WorkItemResponse.model_validate(item)
    response.effective_status = svc.derive_effective_status(item)
    if meeting:
        from investhome_api.schemas.work_item import MeetingRecordResponse

        response.meeting = MeetingRecordResponse.model_validate(meeting)
    if follow_up:
        from investhome_api.schemas.work_item import FollowUpRecordResponse

        response.follow_up = FollowUpRecordResponse.model_validate(follow_up)
    if participants:
        from investhome_api.schemas.work_item import WorkItemParticipantResponse

        response.participants = [WorkItemParticipantResponse.model_validate(p) for p in participants]
    return response


@router.get("/dashboard/kpis", response_model=WorkDashboardKpisResponse)
def get_dashboard_kpis(
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkDashboardKpisResponse:
    return WorkDashboardKpisResponse(**svc.build_dashboard_kpis(db, user))


@router.get("/views/{view_name}", response_model=WorkItemListResponse)
def get_view(
    view_name: str,
    limit: int = Query(default=100, ge=1, le=500),
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkItemListResponse:
    try:
        items = svc.list_by_view(db, user, view_name, limit=limit)
    except svc.WorkItemError as exc:
        raise _handle_error(exc) from exc
    return WorkItemListResponse(
        items=[_enrich(db, item) for item in items],
        total=len(items),
        offset=0,
        limit=limit,
    )


@router.get("/calendar", response_model=list[CalendarEventResponse])
def get_calendar_events(
    from_at: datetime | None = Query(default=None),
    to_at: datetime | None = Query(default=None),
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> list[CalendarEventResponse]:
    items, _ = svc.list_work_items(
        db,
        user,
        due_from=from_at,
        due_to=to_at,
        limit=500,
        sort_by="start_at",
    )
    return [
        CalendarEventResponse(
            id=item.id,
            title=item.title,
            work_item_type=item.work_item_type,
            status=svc.derive_effective_status(item),
            start_at=item.start_at or item.due_at,
            due_at=item.due_at,
            assigned_user_id=item.assigned_user_id,
            is_private=item.is_private,
        )
        for item in items
    ]


@router.get("/items", response_model=WorkItemListResponse)
def list_items(
    search: str | None = Query(default=None, max_length=255),
    assigned_user_id: UUID | None = Query(default=None),
    created_by_user_id: UUID | None = Query(default=None),
    work_item_type: WorkItemType | None = Query(default=None),
    status: WorkItemStatus | None = Query(default=None),
    priority: WorkItemPriority | None = Query(default=None),
    due_from: datetime | None = Query(default=None),
    due_to: datetime | None = Query(default=None),
    lead_id: UUID | None = Query(default=None),
    opportunity_id: UUID | None = Query(default=None),
    party_id: UUID | None = Query(default=None),
    proposal_id: UUID | None = Query(default=None),
    is_private: bool | None = Query(default=None),
    include_archived: bool = Query(default=False),
    sort_by: str = Query(default="due_at", max_length=50),
    sort_dir: str = Query(default="asc", pattern="^(asc|desc)$"),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkItemListResponse:
    items, total = svc.list_work_items(
        db,
        user,
        search=search,
        assigned_user_id=assigned_user_id,
        created_by_user_id=created_by_user_id,
        work_item_type=work_item_type,
        status=status,
        priority=priority,
        due_from=due_from,
        due_to=due_to,
        lead_id=lead_id,
        opportunity_id=opportunity_id,
        party_id=party_id,
        proposal_id=proposal_id,
        is_private=is_private,
        include_archived=include_archived,
        sort_by=sort_by,
        sort_dir=sort_dir,
        offset=offset,
        limit=limit,
    )
    return WorkItemListResponse(
        items=[_enrich(db, item) for item in items],
        total=total,
        offset=offset,
        limit=limit,
    )


@router.get("/items/{work_item_id}", response_model=WorkItemResponse)
def get_item(
    work_item_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=user)
    except svc.WorkItemError as exc:
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items", response_model=WorkItemResponse, status_code=status.HTTP_201_CREATED)
def create_item(
    payload: WorkItemCreate,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "create")),
) -> WorkItemResponse:
    try:
        item = svc.create_work_item(db, payload.model_dump(), actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.patch("/items/{work_item_id}", response_model=WorkItemResponse)
def update_item(
    work_item_id: UUID,
    payload: WorkItemUpdate,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "update")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        item = svc.update_work_item(
            db,
            item,
            payload.model_dump(exclude_unset=True),
            actor=actor,
            request=request,
        )
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items/{work_item_id}/status", response_model=WorkItemResponse)
def change_status(
    work_item_id: UUID,
    payload: StatusChangeRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "update")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        item = svc.change_status(
            db,
            item,
            payload.status,
            actor=actor,
            reason=payload.reason,
            request=request,
        )
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items/{work_item_id}/complete", response_model=WorkItemResponse)
def complete_item(
    work_item_id: UUID,
    payload: CompleteWorkItemRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "complete")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        item = svc.complete_work_item(
            db,
            item,
            actor=actor,
            outcome=payload.outcome,
            next_follow_up=payload.next_follow_up,
            request=request,
        )
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items/{work_item_id}/cancel", response_model=WorkItemResponse)
def cancel_item(
    work_item_id: UUID,
    payload: StatusChangeRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "cancel")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        item = svc.cancel_work_item(db, item, actor=actor, reason=payload.reason, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items/{work_item_id}/reschedule", response_model=WorkItemResponse)
def reschedule_item(
    work_item_id: UUID,
    payload: RescheduleRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "update")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        item = svc.reschedule_work_item(db, item, due_at=payload.due_at, actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.delete("/items/{work_item_id}", response_model=WorkItemResponse)
def archive_item(
    work_item_id: UUID,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "archive")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor, include_archived=True)
        item = svc.archive_work_item(db, item, actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/items/{work_item_id}/restore", response_model=WorkItemResponse)
def restore_item(
    work_item_id: UUID,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("work", "archive")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor, include_archived=True)
        item = svc.restore_work_item(db, item, actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/meetings", response_model=WorkItemResponse, status_code=status.HTTP_201_CREATED)
def create_meeting(
    payload: MeetingCreate,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("sales", "manage_meetings")),
) -> WorkItemResponse:
    try:
        item, _ = svc.create_meeting(db, payload.model_dump(), actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.patch("/meetings/{work_item_id}", response_model=WorkItemResponse)
def update_meeting(
    work_item_id: UUID,
    payload: MeetingUpdate,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("sales", "manage_meetings")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        meeting = db.scalar(select(MeetingRecord).where(MeetingRecord.work_item_id == work_item_id))
        if meeting is None:
            raise svc.WorkItemError("work.errors.meeting_not_found", status_code=404)
        svc.update_meeting(db, item, meeting, payload.model_dump(exclude_unset=True), actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/meetings/{work_item_id}/complete", response_model=WorkItemResponse)
def complete_meeting(
    work_item_id: UUID,
    payload: MeetingCompleteRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("sales", "manage_meetings")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        meeting = db.scalar(select(MeetingRecord).where(MeetingRecord.work_item_id == work_item_id))
        if meeting is None:
            raise svc.WorkItemError("work.errors.meeting_not_found", status_code=404)
        svc.complete_meeting(
            db,
            item,
            meeting,
            actor=actor,
            data=payload.model_dump(exclude_unset=True),
            request=request,
        )
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/follow-ups", response_model=WorkItemResponse, status_code=status.HTTP_201_CREATED)
def create_follow_up(
    payload: FollowUpCreate,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("sales", "manage_followups")),
) -> WorkItemResponse:
    try:
        item, _ = svc.create_follow_up(db, payload.model_dump(), actor=actor, request=request)
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.post("/follow-ups/{work_item_id}/complete", response_model=WorkItemResponse)
def complete_follow_up(
    work_item_id: UUID,
    payload: FollowUpCompleteRequest,
    request: Request,
    db: Session = Depends(get_db),
    actor: User = Depends(require_permission("sales", "manage_followups")),
) -> WorkItemResponse:
    try:
        item = svc.get_work_item_or_raise(db, work_item_id, user=actor)
        record = db.scalar(select(FollowUpRecord).where(FollowUpRecord.work_item_id == work_item_id))
        if record is None:
            raise svc.WorkItemError("work.errors.follow_up_not_found", status_code=404)
        svc.complete_follow_up(
            db,
            item,
            record,
            actor=actor,
            data=payload.model_dump(exclude_unset=True),
            request=request,
        )
        db.commit()
        db.refresh(item)
    except svc.WorkItemError as exc:
        db.rollback()
        raise _handle_error(exc) from exc
    return _enrich(db, item)


@router.get("/leads/{lead_id}/items", response_model=WorkItemListResponse)
def list_lead_items(
    lead_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkItemListResponse:
    items = svc.list_for_lead(db, lead_id, user)
    return WorkItemListResponse(
        items=[_enrich(db, item) for item in items],
        total=len(items),
        offset=0,
        limit=len(items),
    )


@router.get("/opportunities/{opportunity_id}/items", response_model=WorkItemListResponse)
def list_opportunity_items(
    opportunity_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission("work", "view")),
) -> WorkItemListResponse:
    items = svc.list_for_opportunity(db, opportunity_id, user)
    return WorkItemListResponse(
        items=[_enrich(db, item) for item in items],
        total=len(items),
        offset=0,
        limit=len(items),
    )
